promoter/image/ratelimit/roundtripper.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratelimit
+
+import (
+	"context"
+	"net/http"
+
+	"golang.org/x/time/rate"
+)
+
+const burst = 1
+
+// RoundTripper wraps an http.RoundTripper with rate limiting
+type RoundTripper struct {
+	rateLimiter  *rate.Limiter
+	roundTripper http.RoundTripper
+}
+
+var _ http.RoundTripper = &RoundTripper{}
+
+func NewRoundTripper(limit rate.Limit) *RoundTripper {
+	return &RoundTripper{
+		rateLimiter:  rate.NewLimiter(limit, burst),
+		roundTripper: http.DefaultTransport,
+	}
+}
+
+func (rt *RoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	// only rate limit read type calls, not writes
+	if r.Method == http.MethodGet || r.Method == http.MethodHead ||
+		r.Method == http.MethodPut || r.Method == http.MethodPost {
```

Comment:
Put / Post should be write calls?

Reply:
I removed the comment but I was wondering if we should sniff at all :thinking:

Reply:
Don't we only want to limit read calls? Put / Post should count as Write and have a different rate limit (none atm).

Reply:
Ah gotcha I thought they had the same limit, reverted to the original.

Reply:
aha. docs: https://cloud.google.com/artifact-registry/quotas

> 60000 requests per minute in each region or multi-region.
> 18000 write or delete requests per minute in each region or multi-region.

these are the shared baselines that can be raised by request (and we have raised some of them), then we have the per-user/per-ip read quota configured but not the per-user write quota (since kpromo should be the _only_ thing performing writes/deletes anyhow)

Reply:
https://github.com/kubernetes/registry.k8s.io/issues/153#issuecomment-1452753165
> Requests per project per user per minute per user
> Write requests per project per user per minute per user

https://github.com/kubernetes/registry.k8s.io/issues/153#issuecomment-1460913153
> Requests per project per user per minute per user  set to 5000.